[ROCm] enable HIPMallocAsyncAllocator

ethanwee1 opened the pull request:
cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

ethanwee1 committed:
[ROCm] enable HIPMallocAsyncAllocator

jeffdaily commented:
@pytorchbot merge

jeffdaily commented:
@pytorchbot merge -f "1 unrelated failure, all other rocm and cuda CI is passing"

albanD commented:
@ZainRizvi is this a difference in compiler setting between internal and regular build? Should this be fixed? cc @malfet in case you know

jeffdaily committed:
fix D71214814

Some CUDA/HIP APIs missing return value check. HIP header declared the
functions with 'nodiscard' which resulted in an error.

jeffdaily commented:
> @pytorchbot revert -c ghfirst -m "Sorry but this is breaking internally. @albanD, might you be able to help get this PR landed? See D71214814 for more details on the failure. To validate the fixes internally, you can follow the instructions here: https://fburl.com/fixing-ghfirst-reverts"
> 
> The error is a compilation error:
> 
> ```
> .../hip/HIPMallocAsyncAllocator.cpp:508:9: error: ignoring return value of function declared with 'nodiscard' attribute [-Werror,-Wunused-result]
>   508 |         hipDeviceGetDefaultMemPool(&mempool, dev);
>       |         ^~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~
> .../hip/HIPMallocAsyncAllocator.cpp:509:9: error: ignoring return value of function declared with 'nodiscard' attribute [-Werror,-Wunused-result]
>   509 |         hipDeviceSynchronize();
>       |         ^~~~~~~~~~~~~~~~~~~~
> .../hip/HIPMallocAsyncAllocator.cpp:510:9: error: ignoring return value of function declared with 'nodiscard' attribute [-Werror,-Wunused-result]
>   510 |         hipMemPoolTrimTo(mempool, 0);
>       |         ^~~~~~~~~~~~~~~~ ~~~~~~~~~~
> .../hip/HIPMallocAsyncAllocator.cpp:586:9: error: ignoring return value of function declared with 'nodiscard' attribute [-Werror,-Wunused-result]
>   586 |         hipFreeAsync(dummy, stream);
>       |         ^~~~~~~~~~~~ ~~~~~~~~~~~~~
> 4 errors generated.
> ```

I think I fixed this with latest commit.  emptyCache() had 3 calls to CUDA/HIP APIs that were missing C10_CUDA_CHECK.  HIP header declares these functions with nodiscard but CUDA must not have been as strict.

jeffdaily commented:
@ZainRizvi problem should be fixed now.  Can I reland without Meta importing it?  Or should I wait for feedback?

CC @albanD / @malfet

jeffdaily commented:
@pytorchbot merge

jeffdaily commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

@jeffdaily, sorry, it looks like one error remains:
```
buck-out/v2/gen/fbcode/bc717eae0fe2276a/caffe2/c10/cuda/__fb_c10_hipify_gen__/out/c10/hip/HIPMallocAsyncAllocator.cpp:586:9: error: ignoring return value of function declared with 'nodiscard' attribute [-Werror,-Wunused-result]
  586 |         hipFreeAsync(dummy, stream);
      |         ^~~~~~~~~~~~ ~~~~~~~~~~~~~
1 error generated.
```

jeffdaily committed:
one more cuda API missing return code check

jeffdaily commented:
@izaitsevfb fixed.  Should we get a Meta person to import this time?  I didn't like waiting, but I also don't like causing problems and having this back and forth churn.

izaitsevfb commented:
> @izaitsevfb fixed. Should we get a Meta person to import this time? I didn't like waiting, but I also don't like causing problems and having this back and forth churn.

thanks!

let me import it now and run the tests before merge

izaitsevfb commented:
preliminary, the fix works, I'm going to run some more tests just to be sure and then merge the PR

thanks for the help, @jeffdaily!

izaitsevfb commented:
@pytorchbot merge -f "reland with a trivial change, full CI is not needed"

842974287 commented:
@jeffdaily It seems like internally we are failing this [check](https://github.com/pytorch/pytorch/blob/5a7588f1832a840285ed29b039f01b9031570e5c/aten/src/ATen/cuda/CachingHostAllocator.cpp#L196) if we set "pinned_use_cuda_host_register:False". Wondering if anyone from your end can help take a look. Thanks!

jeffdaily commented:
@842974287 I've messaged you directly.
